zoo/repos/views.py

```diff
@@ -14,16 +24,61 @@ class RepoList(ListView):
 
 
 def repo_details(request, provider, repo_id):
-
-    if repo_id is None:
-        return JsonResponse({"error": "repository id not specified"}, status=400)
-    if provider is None:
-        return JsonResponse({"error": "provider is not specified"}, status=400)
-
     scm_module = get_scm_module(provider)
 
     try:
         details = scm_module.get_project_details(repo_id)
     except RepositoryNotFoundError:
         raise Http404(f"Project {repo_id} not found")
+
     return JsonResponse(details)
+
+
+def _parse_file(path, base=None):
+    try:
+        parser = ResolvingParser(str(path))
+        return parser.specification
+    except (
+        ValidationError,
+        ResolutionError,
+        ComposerError,
+        ScannerError,
+        AttributeError,
+    ) as err:
+        log.info(
+            "repos.views.openapi.invalid", path=str(path.relative_to(base)), error=err
+        )
+
+
+def openapi_definition(request, provider, repo_id):
+    try:
+        repository = Repository.objects.get(remote_id=repo_id, provider=provider)
+    except Repository.DoesNotExist:
+        return JsonResponse({"error": "repository not found"}, status=400)
+
+    redis_conn = redis.get_connection()
+    key = f"openapi-{provider}-{repo_id}"
+
+    if redis_conn.exists(key) and "force" not in request.GET:
+        return JsonResponse(json.loads(redis_conn.get(key)), safe=False)
+
+    specs = []
+
+    with tempfile.TemporaryDirectory() as repo_dir:
+        try:
+            repo_path = download_repository(repository, repo_dir)
+        except (MissingFilesError, RepositoryNotFoundError) as err:
+            log.info("repos.views.openapi.git_error", repo=repository, error=err)
+            return JsonResponse({"error": "error downloading repository"}, status=400)
```

Comment:
I'm not sure about the status here as well. 400 means it's a client error, but it doesn't seem to be the case here.

Reply:
Also agree